app/assets/src/components/visualizations/Histogram.js

```diff
@@ -107,26 +130,74 @@ export default class Histogram {
       }
     }
 
-    let x = scaleLinear()
-      .domain([min(mins), max(maxs)])
-      .nice()
-      .range([this.margins.left, this.size.width - this.margins.right]);
+    return [min(mins), max(maxs)];
+  };
+
+  getBins = x => {
+    if (this.options.skipBins) {
+      return this.data;
+    }
 
     let bins = [];
     for (let i = 0; i < this.data.length; i++) {
       bins[i] = histogram()
         .domain(x.domain())
         .thresholds(x.ticks(20))(this.data[i]);
     }
+    return bins;
+  };
+
+  getBarWidth = () => {
+    if (this.options.skipBins) {
+      return (
+        (this.size.width - this.margins.right - this.margins.left) /
+        max(map(seriesData => seriesData.length, this.data))
+      );
+    }
+
+    return (
+      (this.size.width - this.margins.right - this.margins.left) /
+      (this.data.length * 20)
+    );
+  };
+
+  getColors = () => {
+    if (this.options.colors) {
+      return this.options.colors;
+    }
+
+    return new CategoricalColormap().getNScale(this.data.length + 1);
+  };
+
+  getBarOpacity = () => {
+    return this.options.barOpacity || 0.8;
+  };
+
+  update() {
+    if (!this.data) return;
+
+    let colors = this.getColors();
+
+    const domain = this.getDomain();
+
+    let x = scaleLinear()
+      .domain(domain)
+      .nice()
+      .range([this.margins.left, this.size.width - this.margins.right]);
+
+    const bins = this.getBins(x);
 
     let y = scaleLinear()
       .domain([0, max(bins.map(seriesBins => max(seriesBins, d => d.length)))])
       .nice()
       .range([this.size.height - this.margins.bottom, this.margins.top]);
 
-    let barWidth =
-      (this.size.width - this.margins.right - this.margins.left) /
-      (this.data.length * 20);
+    this.svg.append("g").call(this.xAxis.bind(this), x);
```

Comment:
I know this is not new, but we might need to change this in the future. 
As is, and with previous version, we can only call update once. If we need to handle data changes within the same visualization, we will need to move these appends to where we create the svg.

Do you foresee having to change visualization data? Or would you just recreate the histogram?

Reply:
Will refactor in follow-up PR.